rmf_task/include/rmf_task/Request.hpp

```diff
@@ -30,10 +30,11 @@
 
 namespace rmf_task {
 
-/// Implement this for new type of requests.
+//==============================================================================
 class Request
 {
 public:
+  /// Implement this for new type of requests.
   class Description
```

Comment:
While we're refactoring the API, I have a suggestion that would add another layer of complexity to the API, but I think it would help make it more sound, consistent, and stable.

Right now there's a risk that a task description won't match up correctly with the `Configuration` given to the task planner. It's possible that the `motion_sink`, `ambient_sink`, `planner`, etc could get jumbled up where different values are given to the task requests than to the task planner.

To guarantee consistency, we could change the API into this:

```C++
class Request
{
public:

  class Model
  {
  public:
    virtual std::optional<Estimate> estimate_finish(
      const agv::State& initial_state,
      const agv::Constraints& task_planning_constraints,
      const std::shared_ptr<EstimateCache> estimate_cache) const = 0;

    virtual rmf_traffic::Duration invariant_duration() const = 0;

    virtual ~Model();
  };

  class Description
  {
  public:
    virtual std::shared_ptr<Model> make_model(
      rmf_traffic::Time earliest_start_time,
      const Parameters& parameters) const = 0;

    virtual ~Description();
  };

  /* ... The rest of Request remains unchanged ... */
};
```

When the task planner takes in its vector of `requests`, it would then pass the `parameters` from its `configuration` into the `Description::make_model(~)` of each request to get the `Request::Model` instance that it can then use to calculate the estimates.

The APIs for producing tasks would be able to simplify like this:

```C++
class Clean
{
public:

  // Note: the Model class definition for Clean doesn't necessarily
  // need to be exposed to the public API.
  class Model;

  class Description : public Request::Description
  {
  public:
    static DescriptionPtr make(
      std::size_t start_waypoint,
      std::size_t end_waypoint,
      const rmf_traffic::Trajectory& cleaning_path);

    std::shared_ptr<Request::Model> make_model(
      rmf_traffic::Time earliest_start_time,
      const Parameters& parameters) const final;

    std::size_t start_waypoint() const;
    std::size_t end_waypoint() const;
    rmf_traffic::agv::Planner::Start location_after_clean(
      rmf_traffic::agv::Planner::Start start) const;

    class Implementation;
  private:
    Description();
    rmf_utils::impl_ptr<Implementation> _pimpl;
  };

  static ConstRequestPtr make(
    std::size_t start_waypoint,
    std::size_t end_waypoint,
    const rmf_traffic::Trajectory& cleaning_path,
    const std::string& id,
    rmf_traffic::Time earliest_start_time,
    ConstPriorityPtr priority = nullptr);
};
```

This should help satisfy the single-source-of-truth principle and separate the concerns of the classes. `Description` would only be concerned with the generic description of the type of task, unconcerned with the properties of who/what is performing the task, while `Model` would be the tool used by the planner for actually calculating the outcome of performing the task in specific conditions for a specific robot.

Changing the API like this would also help in the future if we decide to offer a task planner API that supports planning across multiple different types of robot platforms simultaneously, because now the `Task::Description` instances are no longer tied to specific robot parameters.